packages/compiler/loader.mjs

```diff
@@ -0,0 +1,34 @@
+import { URL, pathToFileURL, fileURLToPath } from "url";
+import { cwd } from "process";
+import compiler from "@marko/compiler";
+
+const baseURL = pathToFileURL(`${cwd()}/`).href;
+const extensionRegex = /\.marko$/;
+const format = { format: "module" };
+
+export function resolve(specifier, context, defaultResolve) {
+  const { parentURL = baseURL } = context;
+  return extensionRegex.test(specifier)
+    ? {
+        url: new URL(specifier, parentURL).href
+      }
+    : defaultResolve(specifier, context, defaultResolve);
+}
+
+export function getFormat(url, context, defaultGetFormat) {
+  return extensionRegex.test(url)
+    ? format
+    : defaultGetFormat(url, context, defaultGetFormat);
+}
```

Comment:
Heads up that the loader hooks API [has changed](https://nodejs.org/docs/latest-v16.12.0/api/esm.html#hooks) in LTS 16.12.0+ along with 17.x, which will presumably provide the initial API for 18.x. The change eliminates `getFormat`, moving its responsibility to the two remaining functions[^1]:

- `resolve`, as a hint
- `load`, which can use that hint but has the final say

You'll likely want to call this in `resolve` (assuming you need to support 14.x), with the fallback deferring to the result from `defaultResolve`.

[^1]: besides `globalPreload`, which is intended for setup not the dynamic runtime